docs/discussion/Cosign Upgrade 2024.md

```diff
@@ -0,0 +1,468 @@
+        - ~~NOT trivial~~
+        - ~~Require a new http cache server and clients~~
+    - ~~Cache provider initialization in cosign plugin implementation~~
+    - ~~One advantage is that not only will cert + key sharing be solved but also oras store credentials can easily shared which is currently not possible for external plugins~~
+- Extend Inline Certificate Provider
+  - add new `type` field to spec: "key" or "certificate". If not provided, default to "certificate" for backwards compatability. 
+- Extend Azure Key Vault Certificate Provider
+  - use `GetKeys` API to fetch keys from configured key vault
+  - add new `keys` field to spec
+- Multitenancy considerations
+  - Need a separate keys map for namespaced and global certificate store keys
+
+### Option 2: Renaming CertificateStore to KeyManagementSystem
+- Same as Option 1 except we introduce a new CRD that will replace `CertificateStore` due to potential naming confusion
+- Is the existing name confusing enough to warrant changing to KeyManagementSystem?
+- How would we support CRD name change?
+  - Introduce new `KeyManagementSystem` CRD resource and then deprecate `CertificateStore`
+
+### Option 3: Introduce new CRD `KeyStore` separate from `CertificateStore`
+
+All of the same considerations as Option 1. It will be a completely new CRD with an inline and AKV provider equivalent.
+
+Pros: 
+- More intuitive naming: embedding 'key' specification in cert store might be confusing
+- No breaking changes or backwards compatability requirements
+Cons:
+- Yet one more resource type user must be concerned to configure
+- Providers will share almost the exact same implementation as certificate stores
+
+### How do we support CLI scenario?
+  - `CertificateStore` is only supported in K8s scenarios as CRD. There is no CLI equivalent.
+  - Ideally, `CertificateStore` should be supported at CLI too. However, even with CLI mode, a local directory for keys should be an option user can specify rather than just in-line and AKV.
+  - Cosign plugin will handle directory configuration and key reading within plugin implementation. This should be complimentary to any certs provided as plugin input.
+  - Should we have a separate local directory cert store provider that gets configured like other cert stores or should we scope it only to the cosign plugin?
+    - In K8s scenarios, it will not be an encouraged pattern to specify local directory. This is why implementing directory cert reading as a cert store provider may not make  sense.
+    - However, if a user decides to do a mixture of cert stores + local directory then they would need to specify in 2 different sections (cosign plugin config + cert store)
+    - **UPDATE 1/18/24**: We will mirror notation experience with a `verificationKeys` field in the cosign verifier config
+
+## Support Multiple keys for cosign verification
+
+Currently, there is only support for a single key per cosign verifier. 
```

Comment:
TODO: open up an issue for revocation key list?

Reply:
who will own the management of revocation keys? Ratify or third-party service?

Reply:
I am also working on this if only key pairs are used, no certificates. What will be the expiry and revocation experience. 

To clarify the difference between Cosign and Notary Project. The key for cosign verification, it means public key. For notary project, we use root CA certificate, not public key. It's different. Cosign does support verification using certificates, but it is not in the scope of cosign support yet. I will bring more clarification on this soon.

Reply:
Thanks @yizha1 for context. So in public key scenario, if user wants to revoke public key, they would simply just delete the key in their Key Vault? There's no concept of expiry like in certificates?

Reply:
If users manage key pairs in AKV, and the key is compromised, users need to rotate the key, so there will be a new version of the key. If we always retrieve the latest version of the key, then there should be a way to gracefully remove the previous version of the key. There is no concept of key expire, but AKV allows uses to set expiration date, but after expiration, users need to disable the version of key or delete the key totally, then this should be detectable via API, so the question I think is more about how to gracefully remove the old version of key in what context. We can discuss it further in the community meeting and document the behavior.